{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collagen Protein Visualisation With Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of contents:<br>\n",
    "1. [Attributes of Biopython](#1.-Attributes-of-Biopython)\n",
    "2. [Understand FASTA file format](#2.-Understanding-FASTA-file-format)\n",
    "3. [Sequence manipulation using Biopython](#3.-Sequence-manipulation-using-Biopython)\n",
    "4. [Transcription & Translation studies](#4.-Transcription-&-Translation-Studies)\n",
    "5. [Perform Basic Local Alignment using NCBI-BLAST](#5.-Basic-Local-Alignment-Using-NCBI-BLAST)\n",
    "6. [Reading PDB file](#6.-Reading-PDB-file)\n",
    "7. [Visualizing Collagen Protein structure](#7.-Visualizing-Collagen-Protein-structure)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\" style=\"Font-size:16px\">\n",
    "    \n",
    "- Set of **tools** for **Biological computation**<br>\n",
    "- Written in **Python**<br>\n",
    "- Distributed **collaborative effort** to develop Python libraries and applications<br>\n",
    "- Address the needs of current and future work in **Bioinformatics**<br>\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Applications"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\" style=\"Font-size:15px\">\n",
    "    \n",
    "1. **Sequence Analysis** (DNA/RNA/Protein)<br>\n",
    "2. **Transcription** & **Translation studies**<br>\n",
    "3. Accessing **Bioinformatics Databases**<br>\n",
    "a. NCBI<br>\n",
    "b. PDB<br>\n",
    "\n",
    "4. **3D Structure** Analysis\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MODULES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (1.84)\n",
      "Requirement already satisfied: numpy in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from biopython) (1.26.4)\n",
      "Requirement already satisfied: nglview in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (3.1.2)\n",
      "Requirement already satisfied: ipywidgets>=8 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nglview) (8.1.5)\n",
      "Requirement already satisfied: notebook>=7 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nglview) (7.0.8)\n",
      "Requirement already satisfied: jupyterlab-widgets in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nglview) (3.0.13)\n",
      "Requirement already satisfied: numpy in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nglview) (1.26.4)\n",
      "Requirement already satisfied: comm>=0.1.3 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipywidgets>=8->nglview) (0.2.1)\n",
      "Requirement already satisfied: ipython>=6.1.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipywidgets>=8->nglview) (8.25.0)\n",
      "Requirement already satisfied: traitlets>=4.3.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipywidgets>=8->nglview) (5.14.3)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0.12 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipywidgets>=8->nglview) (4.0.13)\n",
      "Requirement already satisfied: jupyter-server<3,>=2.4.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from notebook>=7->nglview) (2.14.1)\n",
      "Requirement already satisfied: jupyterlab-server<3,>=2.22.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from notebook>=7->nglview) (2.25.1)\n",
      "Requirement already satisfied: jupyterlab<4.1,>=4.0.2 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from notebook>=7->nglview) (4.0.11)\n",
      "Requirement already satisfied: notebook-shim<0.3,>=0.2 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from notebook>=7->nglview) (0.2.3)\n",
      "Requirement already satisfied: tornado>=6.2.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from notebook>=7->nglview) (6.4.1)\n",
      "Requirement already satisfied: decorator in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipython>=6.1.0->ipywidgets>=8->nglview) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipython>=6.1.0->ipywidgets>=8->nglview) (0.18.1)\n",
      "Requirement already satisfied: matplotlib-inline in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipython>=6.1.0->ipywidgets>=8->nglview) (0.1.6)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.41 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipython>=6.1.0->ipywidgets>=8->nglview) (3.0.43)\n",
      "Requirement already satisfied: pygments>=2.4.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipython>=6.1.0->ipywidgets>=8->nglview) (2.15.1)\n",
      "Requirement already satisfied: stack-data in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipython>=6.1.0->ipywidgets>=8->nglview) (0.2.0)\n",
      "Requirement already satisfied: colorama in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipython>=6.1.0->ipywidgets>=8->nglview) (0.4.6)\n",
      "Requirement already satisfied: anyio>=3.1.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (4.2.0)\n",
      "Requirement already satisfied: argon2-cffi>=21.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (21.3.0)\n",
      "Requirement already satisfied: jinja2>=3.0.3 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (3.1.4)\n",
      "Requirement already satisfied: jupyter-client>=7.4.4 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (8.6.0)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (5.7.2)\n",
      "Requirement already satisfied: jupyter-events>=0.9.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.10.0)\n",
      "Requirement already satisfied: jupyter-server-terminals>=0.4.4 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.4.4)\n",
      "Requirement already satisfied: nbconvert>=6.4.4 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (7.10.0)\n",
      "Requirement already satisfied: nbformat>=5.3.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (5.9.2)\n",
      "Requirement already satisfied: overrides>=5.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (7.4.0)\n",
      "Requirement already satisfied: packaging>=22.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (23.2)\n",
      "Requirement already satisfied: prometheus-client>=0.9 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.14.1)\n",
      "Requirement already satisfied: pywinpty>=2.0.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (2.0.10)\n",
      "Requirement already satisfied: pyzmq>=24 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (25.1.2)\n",
      "Requirement already satisfied: send2trash>=1.8.2 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.8.2)\n",
      "Requirement already satisfied: terminado>=0.8.3 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.17.1)\n",
      "Requirement already satisfied: websocket-client>=1.7 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.8.0)\n",
      "Requirement already satisfied: async-lru>=1.0.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyterlab<4.1,>=4.0.2->notebook>=7->nglview) (2.0.4)\n",
      "Requirement already satisfied: ipykernel in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyterlab<4.1,>=4.0.2->notebook>=7->nglview) (6.28.0)\n",
      "Requirement already satisfied: jupyter-lsp>=2.0.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyterlab<4.1,>=4.0.2->notebook>=7->nglview) (2.2.0)\n",
      "Requirement already satisfied: babel>=2.10 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (2.11.0)\n",
      "Requirement already satisfied: json5>=0.9.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (0.9.6)\n",
      "Requirement already satisfied: jsonschema>=4.18.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (4.19.2)\n",
      "Requirement already satisfied: requests>=2.31 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (2.32.2)\n",
      "Requirement already satisfied: idna>=2.8 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from anyio>=3.1.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (3.7)\n",
      "Requirement already satisfied: sniffio>=1.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from anyio>=3.1.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.3.0)\n",
      "Requirement already satisfied: argon2-cffi-bindings in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from argon2-cffi>=21.1->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (21.2.0)\n",
      "Requirement already satisfied: pytz>=2015.7 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from babel>=2.10->jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (2024.1)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jedi>=0.16->ipython>=6.1.0->ipywidgets>=8->nglview) (0.8.3)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jinja2>=3.0.3->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (2.1.3)\n",
      "Requirement already satisfied: attrs>=22.2.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (23.1.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (2023.7.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (0.30.2)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (0.10.6)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-client>=7.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (2.9.0.post0)\n",
      "Requirement already satisfied: platformdirs>=2.5 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-core!=5.0.*,>=4.12->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (3.10.0)\n",
      "Requirement already satisfied: pywin32>=300 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-core!=5.0.*,>=4.12->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (305.1)\n",
      "Requirement already satisfied: python-json-logger>=2.0.4 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (2.0.7)\n",
      "Requirement already satisfied: pyyaml>=5.3 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (6.0.1)\n",
      "Requirement already satisfied: rfc3339-validator in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.1.4)\n",
      "Requirement already satisfied: rfc3986-validator>=0.1.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.1.1)\n",
      "Requirement already satisfied: beautifulsoup4 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (4.12.3)\n",
      "Requirement already satisfied: bleach!=5.0.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (4.1.0)\n",
      "Requirement already satisfied: defusedxml in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.7.1)\n",
      "Requirement already satisfied: jupyterlab-pygments in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.1.2)\n",
      "Requirement already satisfied: mistune<4,>=2.0.3 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (2.0.4)\n",
      "Requirement already satisfied: nbclient>=0.5.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.8.0)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.5.0)\n",
      "Requirement already satisfied: tinycss2 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.2.1)\n",
      "Requirement already satisfied: fastjsonschema in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from nbformat>=5.3.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (2.16.2)\n",
      "Requirement already satisfied: wcwidth in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from prompt-toolkit<3.1.0,>=3.0.41->ipython>=6.1.0->ipywidgets>=8->nglview) (0.2.5)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from requests>=2.31->jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (2.0.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from requests>=2.31->jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (2.2.2)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from requests>=2.31->jupyterlab-server<3,>=2.22.1->notebook>=7->nglview) (2024.7.4)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipykernel->jupyterlab<4.1,>=4.0.2->notebook>=7->nglview) (1.6.7)\n",
      "Requirement already satisfied: nest-asyncio in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipykernel->jupyterlab<4.1,>=4.0.2->notebook>=7->nglview) (1.6.0)\n",
      "Requirement already satisfied: psutil in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from ipykernel->jupyterlab<4.1,>=4.0.2->notebook>=7->nglview) (5.9.0)\n",
      "Requirement already satisfied: executing in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from stack-data->ipython>=6.1.0->ipywidgets>=8->nglview) (0.8.3)\n",
      "Requirement already satisfied: asttokens in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from stack-data->ipython>=6.1.0->ipywidgets>=8->nglview) (2.0.5)\n",
      "Requirement already satisfied: pure-eval in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from stack-data->ipython>=6.1.0->ipywidgets>=8->nglview) (0.2.2)\n",
      "Requirement already satisfied: six>=1.9.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from bleach!=5.0.0->nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.16.0)\n",
      "Requirement already satisfied: webencodings in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from bleach!=5.0.0->nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (0.5.1)\n",
      "Requirement already satisfied: fqdn in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.5.1)\n",
      "Requirement already satisfied: isoduration in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (20.11.0)\n",
      "Requirement already satisfied: jsonpointer>1.13 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (2.1)\n",
      "Requirement already satisfied: uri-template in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.3.0)\n",
      "Requirement already satisfied: webcolors>=1.11 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (24.8.0)\n",
      "Requirement already satisfied: cffi>=1.0.1 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from argon2-cffi-bindings->argon2-cffi>=21.1->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.16.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from beautifulsoup4->nbconvert>=6.4.4->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (2.5)\n",
      "Requirement already satisfied: pycparser in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi>=21.1->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (2.21)\n",
      "Requirement already satisfied: arrow>=0.15.0 in c:\\users\\nirmi\\anaconda3\\lib\\site-packages (from isoduration->jsonschema[format-nongpl]>=4.18.0->jupyter-events>=0.9.0->jupyter-server<3,>=2.4.0->notebook>=7->nglview) (1.2.3)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "580b0becef7b4fa6b0a3ff6f807ac43a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "!pip install biopython\n",
    "!pip install nglview\n",
    "\n",
    "import Bio\n",
    "import pylab\n",
    "import urllib\n",
    "import pandas as pd\n",
    "import nglview as nv\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Blast import NCBIWWW\n",
    "from collections import Counter\n",
    "from Bio.Data import CodonTable\n",
    "from Bio import SeqIO, SearchIO\n",
    "from Bio.PDB import PDBParser, MMCIFParser\n",
    "from Bio.SeqUtils import molecular_weight, gc_fraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Attributes of Biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Align',\n",
       " 'BiopythonDeprecationWarning',\n",
       " 'BiopythonExperimentalWarning',\n",
       " 'BiopythonParserWarning',\n",
       " 'BiopythonWarning',\n",
       " 'Blast',\n",
       " 'Data',\n",
       " 'File',\n",
       " 'GenBank',\n",
       " 'MissingExternalDependencyError',\n",
       " 'MissingPythonDependencyError',\n",
       " 'PDB',\n",
       " 'SVDSuperimposer',\n",
       " 'SearchIO',\n",
       " 'Seq',\n",
       " 'SeqFeature',\n",
       " 'SeqIO',\n",
       " 'SeqRecord',\n",
       " 'SeqUtils',\n",
       " 'Sequencing',\n",
       " 'StreamModeError',\n",
       " 'SwissProt',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__',\n",
       " '__version__',\n",
       " '_parent_dir',\n",
       " '_utils',\n",
       " 'os',\n",
       " 'warnings']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check Attributes of Biopython\n",
    "dir(Bio)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Understanding FASTA file format\n",
    "\n",
    "#### SEQUENCE SOURCE\n",
    "\n",
    "<a href=\"https://www.ncbi.nlm.nih.gov/nuccore/MN908947.3?report=fasta\"> **COLLAGEN SEQUENCE FROM NCBI**<a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1. FASTA Format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2. Reading from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO, SearchIO\n",
    "# Loading FASTA file\n",
    "seq_file_read = SeqIO.read(\"C:/Users/nirmi/HIV2.fna\", \"fasta\")\n",
    "# seq_file_parse = SeqIO.parse(\"sequence.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Bio.SeqRecord.SeqRecord"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(seq_file_read)#,type(seq_file_parse)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Sequence details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'KP890355.1'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list sequence details\n",
    "seq_file_read.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqfromfile = seq_file_read.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: KP890355.1\n",
      "Name: KP890355.1\n",
      "Description: KP890355.1 HIV-2 isolate NWK08 from USA, complete genome\n",
      "Number of features: 0\n",
      "Seq('GCTACCAGGTAGAGCCCCGAGTGGATCCCTGGTAGCACTAGCAGGTGAGCCTGG...CCA')\n"
     ]
    }
   ],
   "source": [
    "# List sequence details from the FASTA file\n",
    "for record in SeqIO.parse(\"C:/Users/nirmi/HIV2.fna\", \"fasta\"):\n",
    "    print(record)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KP890355.1\n"
     ]
    }
   ],
   "source": [
    "print(record.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GCTACCAGGTAGAGCCCCGAGTGGATCCCTGGTAGCACTAGCAGGTGAGCCTGG...CCA')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# store sequence for later analysis\n",
    "seqfromfile = record.seq\n",
    "seqfromfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9480"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# length of sequence\n",
    "len(seqfromfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Molecular Weight: 2944530.0546 Da\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqUtils import molecular_weight\n",
    "\n",
    "# Load your sequence from a FASTA file\n",
    "fasta_path = \"C:/Users/nirmi/HIV2.fna\"\n",
    "record = SeqIO.read(fasta_path, format=\"fasta\")\n",
    "\n",
    "# Calculate molecular weight \n",
    "mol_weight = molecular_weight(record.seq)\n",
    "print(f\"Molecular Weight: {mol_weight} Da\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Sequence manipulation using Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\" style=\"Font-size:16px\">\n",
    "- indexing/slicing<br>\n",
    "- concatination<br>\n",
    "- codon search<br>\n",
    "- GC content<br>\n",
    "- complement<br>\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1. Indexing / Slicing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GCT')"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CODON\n",
    "seqfromfile[0:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2. Concatination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GCTCCA')"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqfromfile[0:3]+seqfromfile[-3:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GCTACCTACCA')"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat = seqfromfile[:6]+seqfromfile[-5:]\n",
    "concat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3. Codon Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqfromfile.find('AGA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4. GC Content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44.324894514767934"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# LOGIC\n",
    "(seqfromfile.count('G') + seqfromfile.count('C'))/(len(seqfromfile)) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC content: 44.324894514767934%\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqUtils import gc_fraction\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Load the sequence from file\n",
    "seq_file_read = SeqIO.read(\"C:/Users/nirmi/HIV2.fna\", \"fasta\")\n",
    "\n",
    "# Calculate GC content\n",
    "gc_content = gc_fraction(seq_file_read.seq) * 100  # Convert fraction to percentage\n",
    "\n",
    "print(f\"GC content: {gc_content}%\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.5. Complement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\" style=\"Font-size:16px\">\n",
    "\n",
    "In **DNA** <br>\n",
    "    **A** Bonds with **T** (DOUBLE BOND)<br>\n",
    "    **G** Bonds with **C** (TRIPLE BOND)<br>\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GCTACC')"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqfromfile[0:6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CGATGG')"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#complement\n",
    "\n",
    "seqfromfile[0:6].complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GGTAGC')"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#reverse complement\n",
    "\n",
    "seqfromfile[0:6].reverse_complement()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Transcription & Translation Studies\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\" style=\"font-size:16px\">\n",
    "DNA > RNA = Transcription\n",
    "</div>\n",
    "<div class=\"alert alert-block alert-warning\" style=\"font-size:16px\">\n",
    "mRNA > amino acid (protein) = Translation\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1. Transcription"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GCUACCAGGU')"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Transcribe\n",
    "\n",
    "mRNA = seqfromfile.transcribe()\n",
    "mRNA[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GCTACCAGGT')"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Back transcribe\n",
    "\n",
    "mRNA.back_transcribe()[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2. Translation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "print(CodonTable.unambiguous_dna_by_id[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATR*SPEWIP')"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_seq = seqfromfile.translate()\n",
    "protein_seq[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3160"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(protein_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9480"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seqfromfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "print(CodonTable.unambiguous_dna_by_id[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('R', 297),\n",
       " ('S', 249),\n",
       " ('G', 228),\n",
       " ('L', 227),\n",
       " ('T', 218),\n",
       " ('K', 206),\n",
       " ('Q', 163),\n",
       " ('E', 159),\n",
       " ('I', 158),\n",
       " ('N', 151)]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Listing the most common amino acids\n",
    "\n",
    "common_amino = Counter(protein_seq)\n",
    "common_amino.most_common(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "del common_amino['*']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjsAAAHWCAYAAACR5EiaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8fJSN1AAAACXBIWXMAAA9hAAAPYQGoP6dpAABN50lEQVR4nO3deVhV5f7//9eWGUUUERBFJMcMLXO2ARBxSMwx7WiOdPLkkKaejsMnxZPHqRxK0ybnnOqUZpmz4EnNnE1zNjFMiVJERUWB9fujH/vrFkSBjeDq+biudV3ue937Xu+1N8rLe00WwzAMAQAAmFSxwi4AAACgIBF2AACAqRF2AACAqRF2AACAqRF2AACAqRF2AACAqRF2AACAqRF2AACAqRF2AACAqRF2gEK2efNm9enTRzVq1FDx4sVVvnx5tW3bVnv27Mm2/969e9WsWTOVKFFCpUqVUocOHfTzzz8/4KrvbcmSJZo+fXq+xggNDVVoaKhd6gHw12XhcRFA4XrhhRd04cIFvfDCC6pZs6Z+//13TZkyRbt379a6devUtGlTa9+jR4+qQYMGeuKJJzR8+HDduHFDo0ePVlJSkvbv36+yZcsW4p7YioyM1KFDhxQXF5fnMQ4fPixJqlmzpp2qAvBXRNgBClliYqJ8fHxs2q5evaoqVaooODhYGzdutLZ37txZMTExOnXqlEqWLClJOnPmjKpWrarXX39dkyZNKrA6r1+/LldXV1kslvvqb4+wAwD2wGEsoJDdGXQkqUSJEqpZs6bi4+OtbWlpafrmm2/UsWNHa9CRpMDAQIWFhWnFihX33JbFYtGAAQP04Ycfqlq1anJxcVHNmjW1bNkym37z58+XxWLR+vXr1adPH5UtW1bu7u5KTU1VRkaGJk+erBo1asjFxUU+Pj7q0aOHzp49a31/aGioVq9erTNnzshisViXTDdv3tS4ceOsY5QtW1a9e/fW77//blPHnYex4uLiZLFY9M4772jq1KkKCgpSiRIl1LhxY+3YseOe+3/t2jUNGzZMQUFBcnV1lZeXl+rVq6elS5fa9Nu9e7eef/55eXl5ydXVVXXq1NFnn32WZbwdO3boqaeekqurq/z9/TVixAh9/PHHslgsNiHPYrEoOjo6y/srVaqkXr162bQlJCSob9++qlChgpydnRUUFKSxY8cqLS0tz5/DDz/8oDZt2qhMmTJydXVV5cqVNXjwYJs+J06cUNeuXeXj4yMXFxc9+uijev/99236ZGRkaNy4capevbrc3NxUqlQp1a5dW+++++5dPnGgaHAs7AIAZJWcnKy9e/faHMI6deqUrl+/rtq1a2fpX7t2bW3YsEE3btyQq6trjmOvWrVKMTEx+ve//63ixYtr1qxZ+tvf/iZHR0d16tTJpm+fPn3UunVrLVq0SCkpKXJyctKrr76qjz76SAMGDFBkZKTi4uL05ptvKjY2Vnv37pW3t7dmzZqlV155RadOncoSwjIyMtS2bVt99913euONN9SkSROdOXNGY8aMUWhoqHbv3i03N7cc9+H9999XjRo1rOcEvfnmm3ruued0+vRpeXp63vV9Q4YM0aJFizRu3DjVqVNHKSkpOnTokC5cuGDtExMTo5YtW6phw4b64IMP5OnpqWXLlqlLly66du2aNZwcPnxY4eHhqlSpkubPny93d3fNmjVLS5YsybH2nCQkJKhBgwYqVqyYRo8ercqVK+v777/XuHHjFBcXp3nz5uX6c1i3bp3atGmjRx99VFOnTlXFihUVFxen9evXW8c5fPiwmjRpoooVK2rKlCny8/PTunXr9Nprr+mPP/7QmDFjJEmTJ09WdHS0/u///k/PPvusbt26paNHj+rSpUt53mfggTAAFDndunUzHB0djd27d1vbtm3bZkgyli5dmqX/+PHjDUnGuXPnchxXkuHm5mYkJCRY29LS0owaNWoYVapUsbbNmzfPkGT06NHD5v1HjhwxJBn9+vWzaf/hhx8MScbIkSOtba1btzYCAwOz1LB06VJDkvHFF1/YtO/atcuQZMyaNcvaFhISYoSEhFhfnz592pBk1KpVy0hLS7O279y5866fze2Cg4ONdu3a5dinRo0aRp06dYxbt27ZtEdGRhrlypUz0tPTDcMwjC5dutz1s5RknD592touyRgzZkyWbQUGBho9e/a0vu7bt69RokQJ48yZMzb93nnnHUOS8dNPP+X6c6hcubJRuXJl4/r163fd5xYtWhgVKlQwkpOTbdoHDBhguLq6GhcvXrR+Bk888cRdxwGKKg5jAUXMm2++qcWLF2vatGmqW7dulvU5nTNzP+fThIeHy9fX1/rawcFBXbp00cmTJ20ORUlSx44dbV7HxMRIUpZDLw0aNNCjjz6qTZs23XP733zzjUqVKqU2bdooLS3NujzxxBPy8/NTbGzsPcdo3bq1HBwcrK8zZ7vOnDmT4/saNGigNWvWaPjw4YqNjdX169dt1p88eVJHjx5Vt27dJMmmvueee07nz5/XsWPHJP35Wdzts8yrb775RmFhYfL397fZdqtWrSRJW7Zssel/r8/h+PHjOnXqlKKiou4643fjxg1t2rRJ7du3l7u7e5Z9vnHjhvXQWIMGDXTgwAH169dP69at0+XLl/O8r8CDRNgBipCxY8dq3Lhx+s9//qMBAwbYrCtTpowk2RxyyXTx4kVZLBaVKlXqntvw8/O7a9udY5crV87mdeb6O9slyd/fP9va7vTbb7/p0qVLcnZ2lpOTk82SkJCgP/74455jZH4WmVxcXCQpS3i503vvvad//etfWrlypcLCwuTl5aV27drpxIkT1tokadiwYVlq69evnyRZ67tw4UKOn2Ve/Pbbb/r666+zbPuxxx6z2Xame30OmedAVahQ4a7bvHDhgtLS0jRjxows233uuedstjtixAi988472rFjh1q1aqUyZcooPDxcu3fvzvM+Aw8C5+wARcTYsWMVHR2t6OhojRw5Msv6ypUry83NTQcPHsyy7uDBg6pSpco9z9eR/jwv5G5td/7yvHOmKHP9+fPns/wCPXfunLy9ve+5fW9vb5UpU0Zr167Ndr2Hh8c9x8ir4sWLa+zYsRo7dqx+++036yxPmzZtdPToUWv9I0aMUIcOHbIdo3r16pL+/Cxy+ixv5+LiotTU1Cztd4ZDb29v1a5dW//5z3+y3ba/v3/OO3iHzFsR3Dljd7vSpUvLwcFB3bt3V//+/bPtExQUJElydHTUkCFDNGTIEF26dEkbN27UyJEj1aJFC8XHx8vd3T1X9QEPCmEHKALeeust64mfmSeD3snR0VFt2rTRl19+qcmTJ1tDwS+//KKYmBi9/vrr97WtTZs26bfffrMefklPT9fy5ctVuXLlHGcAJFlPmP70009Vv359a/uuXbt05MgRjRo1ytrm4uKS7UxLZGSkli1bpvT0dDVs2PC+ai4Ivr6+6tWrlw4cOKDp06fr2rVrql69uqpWraoDBw5o/PjxOb4/LCxMq1atyvazvFOlSpX0448/2rRt3rxZV69etWmLjIzUt99+q8qVK6t06dL53EOpWrVqqly5subOnashQ4ZYZ35u5+7urrCwMO3bt0+1a9eWs7PzfY1dqlQpderUSb/++qsGDx6suLg47oeEIouwAxSyKVOmaPTo0WrZsqVat26d5dLhRo0aWf88duxY1a9fX5GRkTY3FfT29tbQoUPva3ve3t5q2rSp3nzzTevVWEePHs1y+Xl2qlevrldeeUUzZsxQsWLF1KpVK+vVWAEBATaBq1atWvryyy81e/Zs1a1bV8WKFVO9evX04osvavHixXruuec0aNAgNWjQQE5OTjp79qxiYmLUtm1btW/f/j4/vdxp2LChIiMjVbt2bZUuXVpHjhzRokWL1LhxY+usxIcffqhWrVqpRYsW6tWrl8qXL6+LFy/qyJEj2rt3rz7//HNJ0v/93/9p1apVatq0qUaPHi13d3e9//77SklJybLd7t27680339To0aMVEhKiw4cPa+bMmVmuHPv3v/+tDRs2qEmTJnrttddUvXp13bhxQ3Fxcfr222/1wQcf3DOQ3un9999XmzZt1KhRI73++uuqWLGifvnlF61bt06LFy+WJL377rt6+umn9cwzz+jVV19VpUqVdOXKFZ08eVJff/21Nm/eLElq06aNgoODVa9ePZUtW1ZnzpzR9OnTFRgYqKpVq+b6+wAemMI+Qxr4qwsJCTEk3XW50+7du43w8HDD3d3dKFmypNGuXTvj5MmT97UtSUb//v2NWbNmGZUrVzacnJyMGjVqGIsXL7bpl3k11q5du7KMkZ6ebkyaNMmoVq2a4eTkZHh7exsvvfSSER8fb9Pv4sWLRqdOnYxSpUoZFovFZl9u3bplvPPOO8bjjz9uuLq6GiVKlDBq1Khh9O3b1zhx4oTNZ5Pd1Vhvv/12tvuW3RVPtxs+fLhRr149o3Tp0oaLi4vxyCOPGK+//rrxxx9/2PQ7cOCA0blzZ8PHx8dwcnIy/Pz8jKZNmxoffPCBTb9t27YZjRo1MlxcXAw/Pz/jn//8p/HRRx9luRorNTXVeOONN4yAgADDzc3NCAkJMfbv35/laizDMIzff//deO2114ygoCDDycnJ8PLyMurWrWuMGjXKuHr1ap4+h++//95o1aqV4enpabi4uBiVK1c2Xn/9dZs+p0+fNvr06WOUL1/ecHJyMsqWLWs0adLEGDdunLXPlClTjCZNmhje3t6Gs7OzUbFiRSMqKsqIi4vL8XMHCht3UAb+QiwWi/r376+ZM2cWdimmNX/+fPXu3VunT59WpUqVCrscAOJqLAAAYHKEHQAAYGocxgIAAKbGzA4AADA1wg7wkNq3b59CQkLk6ekpi8VifRgkcuell16SxWJRZGRklnVXrlzRa6+9pvLly8vFxUXVqlXT5MmTlZ6efs9xz507p+joaO3fv9+u9R4/flzDhg1T3bp1VapUKXl5eempp57Sf//732z7r1u3Tk899ZTc3Nzk6empNm3a6KeffrLpk/kU9bstLVu2tOs+AA8a99kBHlJ9+vRRSkqKli1bptKlS3PlTx6sXr1aK1euVMmSJbOsS0tLU0REhI4fP6633npL1apV09q1azV8+HCdPXtW7733Xo5jnzt3TmPHjlWlSpX0xBNP2K3m9evXa/Xq1erevbvq16+vtLQ0LV++XC+88ILGjh2r0aNHW/t+9dVXat++vdq2basvvvhCycnJGjt2rJ555hnt2rVLlStXlvTn4z++//77LNtauXKlJk2aVGD3PQIemMK98h1AXjk6OhqvvvrqPftdu3bNyMjIeAAVPVwuXbpklC9f3pg6daoRGBhotG7d2mb93Z7O/sorrxjFihUzjh49muP4mU9xnzdvnl3r/v3337P9Plu3bm24u7sbN27csLZVr17dqF27tk3/uLg4w9nZ2ejates9txUaGmq4u7tneRo68LDhMBbwkJk/f74sFovS0tI0e/Zs66GG29etX79effr0UdmyZeXu7q7U1FQZhqHJkycrMDBQrq6uevLJJ7VmzRqFhoYqNDQ0y/hxcXE2242NjZXFYsnyVPKNGzcqPDxcJUuWlLu7u5566qksTz+Pjo6WxWLRTz/9pL/97W/y9PSUr6+v+vTpo+TkZJu+GRkZmjFjhp544gm5ubmpVKlSatSokVatWiVJioqKkpeXl65du5bls2natKn1oZn3MnToUJUrV06vvfZatuu3bdsmi8VifeJ4psjISGVkZGjFihV3HTs2Ntb6OI3evXtbv6Po6Ghrn1WrVlnv3Ozh4aGIiIhsZ1fu5O3tne3T7Rs0aKBr167p4sWLkv587taxY8fUqlUrm/6BgYEKDg7WypUrczwcd+rUKW3ZskWdO3fOduYLeJgQdoCHTOvWra2/FDt16qTvv/8+yy/JPn36yMnJSYsWLdJ///tfOTk5aezYsfrXv/6liIgIrVy5Uq+++qr+/ve/69ixY3mu5dNPP1Xz5s1VsmRJLViwQJ999pm8vLzUokWLLIFHkjp27Khq1arpiy++0PDhw7VkyZIsz/Tq1auXBg0apPr162v58uVatmyZnn/+eWv4GjRokJKSkrRkyRKb9x0+fFgxMTF3fZjl7TZu3KiFCxfqk08+kYODQ7Z9bt68qWLFisnJycmmPfP5Unc+6+p2Tz75pObNmyfpz8dKZH5HL7/8siRpyZIlatu2rUqWLKmlS5dqzpw5SkpKUmhoqLZu3XrP+rMTExOjsmXLysfHx1r/7fXeuQ/Xrl3TqVOn7jre3LlzZRiGtWbgoVbYU0sA8kb//6Mfbpf5mIcePXrYtCclJRmurq5G+/btbdq3bdtmSLJ5JEPmGLc/7sAwDCMmJsaQZMTExBiGYRgpKSmGl5eX0aZNG5t+6enpxuOPP240aNDA2jZmzBhDkjF58mSbvv369TNcXV2th1n+97//GZKMUaNG5bjvISEhxhNPPGHT9uqrrxolS5Y0rly5kuN7r1y5YlSqVMkYMWKEtS27w1jTp083JBnfffedTfubb75pSDKaN2+e43budhgrPT3d8Pf3N2rVqmWkp6fb1OXj42M0adIkx3Gz8/HHHxuSjHfffddmO15eXkZ4eLhN36SkJMPDw8OQZGzfvj3b8dLS0ozy5csbNWrUyHUtQFHEzA5gQh07drR5/f333+vGjRvq1q2bTXuTJk0UGBiYp21s375dFy9eVM+ePZWWlmZdMjIy1LJlS+3atSvLQzGff/55m9e1a9fWjRs3lJiYKElas2aNJN1zdmbQoEHav3+/tm3bJkm6fPmyFi1apJ49e6pEiRI5vnf48OFycnKyOZE3O926dZOXl5deeeUV/fDDD7p06ZKWLl1qPTG5WLG8/fN57NgxnTt3Tt27d7cZo0SJEurYsaN27NiR7SG6u1mzZo369++vTp06aeDAgdb2YsWKqX///tq0aZPeeustJSYm6uTJk3rppZes499tH9auXatff/1VUVFRedpHoKgh7AAmVK5cOZvXFy5ckCT5+fll6Ztd2/347bffJP15KM3JyclmmTRpkgzDsJ4/kqlMmTI2rzMPsVy/fl2S9Pvvv8vBweGeNbVt21aVKlXS+++/L+nP84xSUlLuGZJ27typWbNmafLkybpx44YuXbqkS5cuKSMjQ2lpabp06ZJSU1Ml/XluzNq1ayX9+eT50qVLa+DAgZo6daokqXz58vf8jLKT+V3c+R1Jkr+/vzIyMpSUlHRfY61bt04dOnRQRESEFi9enOVcntGjR+v111/XuHHj5Ovra30yee/evXPchzlz5sjJyUk9evS47/0CijIuPQdM6M5fepkhIyEhIUvfhIQEm8vWXV1dJcn6Sz/TH3/8YfPa29tbkjRjxgw1atQo2zp8fX1zVXfZsmWVnp6uhISEbMNApsxZi5EjR2rKlCmaNWuWwsPDVb169RzHP3z4sAzDyPZS6vj4eJUuXVrTpk3T4MGDJUn169fX4cOHFRcXp5SUFFWtWlV79uyRJD377LO52rdMmd/F+fPns6w7d+6cihUrptKlS99znHXr1qldu3YKCQnRF198IWdn5yx9HB0dNXXqVP373//W6dOn5e3trXLlyqlFixYKCgpShQoVsrwnMTFR33zzjZ5//nnr+T/Aw46ZHeAvoFGjRnJ1ddXixYtt2rdv364zZ87YtGUGnztPwM28GirTU089pVKlSunw4cOqV69etkt2v4Bzknnl0+zZs+/Z9+WXX5azs7O6deumY8eOacCAAfd8T8uWLRUTE5Nl8fX1VaNGjRQTE6NOnTpleV+lSpX02GOPycnJSVOmTJG/v79eeOGFHLd156xVpurVq6t8+fJasmSJjNue1pOSkqIvvvjCeoVWTtavX6927drp6aef1sqVK7M9Cfl2JUqUUK1atVSuXDnt3btXmzZt0qBBg7Ltu3DhQt26dYtDWDAVZnaAv4DSpUtr2LBhGjdunF5++WW98MILio+PV3R0dJZDRvXr11f16tU1bNgwpaWlqXTp0lqxYkWWq4RKlCihGTNmqGfPnrp48aI6deokHx8f/f777zpw4IB+//33+wott3vmmWfUvXt3jRs3Tr/99psiIyPl4uKiffv2yd3d3eaclFKlSqlHjx6aPXu2AgMD1aZNm3uO7+fnl+0hMldXV5UpU8bmEnxJGjVqlDUk/PLLL5o7d65++OEHrV69Wm5ubjluq3LlynJzc9PixYv16KOPqkSJEvL395e/v78mT56sbt26KTIyUn379lVqaqrefvttXbp0SRMnTsxx3K1bt6pdu3by8/PTyJEjs9yhuWbNmtZLxWNjY7Vr1y7Vrl1bhmFo586dmjRpklq2bHnXcDhnzhwFBASoRYsWOdYBPFQK9/xoAHmlHK7G2rVrV5b+GRkZxoQJE4yAgADD2dnZqF27tvH1118bISEhNldjGYZhHD9+3GjevLlRsmRJo2zZssbAgQON1atX21yNlWnLli1G69atDS8vL8PJyckoX7680bp1a+Pzzz+39sm8Guv333/Ptt7br/xKT083pk2bZgQHBxvOzs6Gp6en0bhxY+Prr7/Osk+xsbGGJGPixIn3+allL7ursQzjzyu8KlasaDg7Oxve3t5Gx44djR9//PG+x126dKlRo0YNw8nJyZBkjBkzxrpu5cqVRsOGDQ1XV1ejePHiRnh4uLFt27Z7jpn5Wd5tuf372bZtm9GwYUOjZMmShouLixEcHGy88847xs2bN7MdO/PqvNGjR9/3PgIPA556DvzFZc5m3HmzwIfB0KFDNXv2bMXHx2c5+RkAMnEYC8BDZ8eOHTp+/LhmzZqlvn37EnQA5IiwA+Chk3kSb2RkpMaNG1fY5QAo4jiMBQAATI1LzwEAgKkRdgAAgKkRdgAAgKlxgrKkjIwMnTt3Th4eHllusw8AAIomwzB05coV+fv75/hwXsKO/nweTUBAQGGXAQAA8iA+Pj7bZ71lIuxI8vDwkPTnh5V5m3UAAFC0Xb58WQEBAdbf43dD2NH/e0J0yZIlCTsAADxk7nUKCicoAwAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyvUsDN79mzVrl3ben+bxo0ba82aNdb1hmEoOjpa/v7+cnNzU2hoqH766SebMVJTUzVw4EB5e3urePHiev7553X27NkHvSsAAKCIKtSwU6FCBU2cOFG7d+/W7t271bRpU7Vt29YaaCZPnqypU6dq5syZ2rVrl/z8/BQREaErV65Yxxg8eLBWrFihZcuWaevWrbp69aoiIyOVnp5eWLsFAACKEIthGEZhF3E7Ly8vvf322+rTp4/8/f01ePBg/etf/5L05yyOr6+vJk2apL59+yo5OVlly5bVokWL1KVLF0n/7zlX3377rVq0aHFf27x8+bI8PT2VnJzMHZQBAHhI3O/v7yJzzk56erqWLVumlJQUNW7cWKdPn1ZCQoKaN29u7ePi4qKQkBBt375dkrRnzx7dunXLpo+/v7+Cg4OtfbKTmpqqy5cv2ywAAMCcCj3sHDx4UCVKlJCLi4v+8Y9/aMWKFapZs6YSEhIkSb6+vjb9fX19resSEhLk7Oys0qVL37VPdiZMmCBPT0/rwhPPAQAwr0IPO9WrV9f+/fu1Y8cOvfrqq+rZs6cOHz5sXX/nw70Mw7jnA7/u1WfEiBFKTk62LvHx8fnbCQAAUGQVethxdnZWlSpVVK9ePU2YMEGPP/643n33Xfn5+UlSlhmaxMRE62yPn5+fbt68qaSkpLv2yY6Li4v1CjCedA4AgLkVeti5k2EYSk1NVVBQkPz8/LRhwwbrups3b2rLli1q0qSJJKlu3bpycnKy6XP+/HkdOnTI2gcAAPy1ORbmxkeOHKlWrVopICBAV65c0bJlyxQbG6u1a9fKYrFo8ODBGj9+vKpWraqqVatq/Pjxcnd3V9euXSVJnp6eioqK0tChQ1WmTBl5eXlp2LBhqlWrlpo1a1aYu/bQqzR8tV3GiZvY2i7jAACQV4Uadn777Td1795d58+fl6enp2rXrq21a9cqIiJCkvTGG2/o+vXr6tevn5KSktSwYUOtX79eHh4e1jGmTZsmR0dHde7cWdevX1d4eLjmz58vBweHwtotAABQhBS5++wUBu6zkxUzOwCAou6hu88OAABAQSDsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyPsAAAAUyvUsDNhwgTVr19fHh4e8vHxUbt27XTs2DGbPr169ZLFYrFZGjVqZNMnNTVVAwcOlLe3t4oXL67nn39eZ8+efZC7AgAAiqhCDTtbtmxR//79tWPHDm3YsEFpaWlq3ry5UlJSbPq1bNlS58+fty7ffvutzfrBgwdrxYoVWrZsmbZu3aqrV68qMjJS6enpD3J3AABAEeRYmBtfu3atzet58+bJx8dHe/bs0bPPPmttd3FxkZ+fX7ZjJCcna86cOVq0aJGaNWsmSfr0008VEBCgjRs3qkWLFlnek5qaqtTUVOvry5cv22N3AABAEVSkztlJTk6WJHl5edm0x8bGysfHR9WqVdPf//53JSYmWtft2bNHt27dUvPmza1t/v7+Cg4O1vbt27PdzoQJE+Tp6WldAgICCmBvAABAUVBkwo5hGBoyZIiefvppBQcHW9tbtWqlxYsXa/PmzZoyZYp27dqlpk2bWmdmEhIS5OzsrNKlS9uM5+vrq4SEhGy3NWLECCUnJ1uX+Pj4gtsxAABQqAr1MNbtBgwYoB9//FFbt261ae/SpYv1z8HBwapXr54CAwO1evVqdejQ4a7jGYYhi8WS7ToXFxe5uLjYp3DkWqXhq+0yTtzE1nYZBwBgbkViZmfgwIFatWqVYmJiVKFChRz7litXToGBgTpx4oQkyc/PTzdv3lRSUpJNv8TERPn6+hZYzQAA4OFQqGHHMAwNGDBAX375pTZv3qygoKB7vufChQuKj49XuXLlJEl169aVk5OTNmzYYO1z/vx5HTp0SE2aNCmw2gEAwMOhUA9j9e/fX0uWLNFXX30lDw8P6zk2np6ecnNz09WrVxUdHa2OHTuqXLlyiouL08iRI+Xt7a327dtb+0ZFRWno0KEqU6aMvLy8NGzYMNWqVct6dRYAAPjrKtSwM3v2bElSaGioTfu8efPUq1cvOTg46ODBg1q4cKEuXbqkcuXKKSwsTMuXL5eHh4e1/7Rp0+To6KjOnTvr+vXrCg8P1/z58+Xg4PAgdwcAABRBhRp2DMPIcb2bm5vWrVt3z3FcXV01Y8YMzZgxw16lAQAAkygSJygDAAAUFMIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcIOAAAwNcfCLgDAw6nS8NV2GSduYmu7jAMAd0PYAQA7IPwBRReHsQAAgKkRdgAAgKkRdgAAgKkRdgAAgKkRdgAAgKkRdgAAgKkRdgAAgKkRdgAAgKkRdgAAgKlxB+WHGHdsBQDg3pjZAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApuZY2AUAAPBXUGn4aruMEzextV3G+SthZgcAAJgaYQcAAJhaoYadCRMmqH79+vLw8JCPj4/atWunY8eO2fQxDEPR0dHy9/eXm5ubQkND9dNPP9n0SU1N1cCBA+Xt7a3ixYvr+eef19mzZx/krgAAgCKqUMPOli1b1L9/f+3YsUMbNmxQWlqamjdvrpSUFGufyZMna+rUqZo5c6Z27dolPz8/RURE6MqVK9Y+gwcP1ooVK7Rs2TJt3bpVV69eVWRkpNLT0wtjtwAAQBFSqCcor1271ub1vHnz5OPjoz179ujZZ5+VYRiaPn26Ro0apQ4dOkiSFixYIF9fXy1ZskR9+/ZVcnKy5syZo0WLFqlZs2aSpE8//VQBAQHauHGjWrRo8cD3CwAAFB1F6pyd5ORkSZKXl5ck6fTp00pISFDz5s2tfVxcXBQSEqLt27dLkvbs2aNbt27Z9PH391dwcLC1z51SU1N1+fJlmwUAAJhTkQk7hmFoyJAhevrppxUcHCxJSkhIkCT5+vra9PX19bWuS0hIkLOzs0qXLn3XPneaMGGCPD09rUtAQIC9dwcAABQRRSbsDBgwQD/++KOWLl2aZZ3FYrF5bRhGlrY75dRnxIgRSk5Oti7x8fF5LxwAABRpRSLsDBw4UKtWrVJMTIwqVKhgbffz85OkLDM0iYmJ1tkePz8/3bx5U0lJSXftcycXFxeVLFnSZgEAAOZUqGHHMAwNGDBAX375pTZv3qygoCCb9UFBQfLz89OGDRusbTdv3tSWLVvUpEkTSVLdunXl5ORk0+f8+fM6dOiQtQ8AAPjrKtSrsfr3768lS5boq6++koeHh3UGx9PTU25ubrJYLBo8eLDGjx+vqlWrqmrVqho/frzc3d3VtWtXa9+oqCgNHTpUZcqUkZeXl4YNG6ZatWpZr84CAAB/XYUadmbPni1JCg0NtWmfN2+eevXqJUl64403dP36dfXr109JSUlq2LCh1q9fLw8PD2v/adOmydHRUZ07d9b169cVHh6u+fPny8HB4UHtCgAAKKIKNewYhnHPPhaLRdHR0YqOjr5rH1dXV82YMUMzZsywY3UAAMAMisQJygAAAAWFsAMAAEyNsAMAAEyNsAMAAEyNsAMAAEyNsAMAAEyNsAMAAEyNsAMAAEyNsAMAAEyNsAMAAEyNsAMAAEyNsAMAAEwt12Hn9OnTBVEHAABAgch12KlSpYrCwsL06aef6saNGwVREwAAgN3kOuwcOHBAderU0dChQ+Xn56e+fftq586dBVEbAABAvuU67AQHB2vq1Kn69ddfNW/ePCUkJOjpp5/WY489pqlTp+r3338viDoBAADyJM8nKDs6Oqp9+/b67LPPNGnSJJ06dUrDhg1ThQoV1KNHD50/f96edQIAAORJnsPO7t271a9fP5UrV05Tp07VsGHDdOrUKW3evFm//vqr2rZta886AQAA8sQxt2+YOnWq5s2bp2PHjum5557TwoUL9dxzz6lYsT9zU1BQkD788EPVqFHD7sUCAADkVq7DzuzZs9WnTx/17t1bfn5+2fapWLGi5syZk+/iAAAA8ivXYefEiRP37OPs7KyePXvmqSAAAAB7yvU5O/PmzdPnn3+epf3zzz/XggUL7FIUAACAveQ67EycOFHe3t5Z2n18fDR+/Hi7FAUAAGAvuQ47Z86cUVBQUJb2wMBA/fLLL3YpCgAAwF5yHXZ8fHz0448/Zmk/cOCAypQpY5eiAAAA7CXXYefFF1/Ua6+9ppiYGKWnpys9PV2bN2/WoEGD9OKLLxZEjQAAAHmW66uxxo0bpzNnzig8PFyOjn++PSMjQz169OCcHQAAUOTkOuw4Oztr+fLleuutt3TgwAG5ubmpVq1aCgwMLIj6AAAA8iXXYSdTtWrVVK1aNXvWAgAAYHe5Djvp6emaP3++Nm3apMTERGVkZNis37x5s92KAwAAyK9ch51BgwZp/vz5at26tYKDg2WxWAqiLgAAALvIddhZtmyZPvvsMz333HMFUQ8AAIBd5frSc2dnZ1WpUqUgagEAALC7XIedoUOH6t1335VhGAVRDwAAgF3l+jDW1q1bFRMTozVr1uixxx6Tk5OTzfovv/zSbsUBAADkV67DTqlSpdS+ffuCqAUAAMDuch125s2bVxB1AAAAFIg83VQwLS1NsbGxOnXqlLp27SoPDw+dO3dOJUuWVIkSJexdIwD8pVUavtou48RNbG2XcYCHTa7DzpkzZ9SyZUv98ssvSk1NVUREhDw8PDR58mTduHFDH3zwQUHUCQAAkCe5vhpr0KBBqlevnpKSkuTm5mZtb9++vTZt2mTX4gAAAPIrT1djbdu2Tc7OzjbtgYGB+vXXX+1WGAAAgD3kOuxkZGQoPT09S/vZs2fl4eFhl6KAosZe50xInDcBAA9arsNORESEpk+fro8++kiSZLFYdPXqVY0ZM4ZHSABFDCe2AkAews60adMUFhammjVr6saNG+ratatOnDghb29vLV26tCBqBAAAyLNchx1/f3/t379fS5cu1d69e5WRkaGoqCh169bN5oRlAACAoiBP99lxc3NTnz591KdPH3vXAwAAYFe5DjsLFy7McX2PHj3yXAwAAIC95TrsDBo0yOb1rVu3dO3aNTk7O8vd3Z2wAwAAipRc31QwKSnJZrl69aqOHTump59+OtcnKP/vf/9TmzZt5O/vL4vFopUrV9qs79WrlywWi83SqFEjmz6pqakaOHCgvL29Vbx4cT3//PM6e/ZsbncLAACYVK7DTnaqVq2qiRMnZpn1uZeUlBQ9/vjjmjlz5l37tGzZUufPn7cu3377rc36wYMHa8WKFVq2bJm2bt2qq1evKjIyMtt7AQEAgL+ePJ2gnB0HBwedO3cuV+9p1aqVWrVqlWMfFxcX+fn5ZbsuOTlZc+bM0aJFi9SsWTNJ0qeffqqAgABt3LhRLVq0yFU9AADAfHIddlatWmXz2jAMnT9/XjNnztRTTz1lt8IyxcbGysfHR6VKlVJISIj+85//yMfHR5K0Z88e3bp1S82bN7f29/f3V3BwsLZv337XsJOamqrU1FTr68uXL9u9bgDAw4cbcZpTrsNOu3btbF5bLBaVLVtWTZs21ZQpU+xVl6Q/Z35eeOEFBQYG6vTp03rzzTfVtGlT7dmzRy4uLkpISJCzs7NKly5t8z5fX18lJCTcddwJEyZo7Nixdq0VAAAUTXl6NtaD0qVLF+ufg4ODVa9ePQUGBmr16tXq0KHDXd9nGIYsFstd148YMUJDhgyxvr58+bICAgLsUzQAAChS7HKC8oNSrlw5BQYG6sSJE5IkPz8/3bx5U0lJSTb9EhMT5evre9dxXFxcVLJkSZsFAACYU65ndm6fEbmXqVOn5nb4HF24cEHx8fEqV66cJKlu3bpycnLShg0b1LlzZ0nS+fPndejQIU2ePNmu2wYAAA+nXIedffv2ae/evUpLS1P16tUlScePH5eDg4OefPJJa7+cDiNlunr1qk6ePGl9ffr0ae3fv19eXl7y8vJSdHS0OnbsqHLlyikuLk4jR46Ut7e32rdvL0ny9PRUVFSUhg4dqjJlysjLy0vDhg1TrVq1rFdnAQCAv7Zch502bdrIw8NDCxYssJ4YnJSUpN69e+uZZ57R0KFD73us3bt3KywszPo6c9aoZ8+emj17tg4ePKiFCxfq0qVLKleunMLCwrR8+XJ5eHhY3zNt2jQ5Ojqqc+fOun79usLDwzV//nw5ODjkdtcAAIAJ5TrsTJkyRevXr7e5Aqp06dIaN26cmjdvnquwExoaKsMw7rp+3bp19xzD1dVVM2bM0IwZM+57uw+SvS5jlLiUEQCAvMj1CcqXL1/Wb7/9lqU9MTFRV65csUtRAAAA9pLrmZ327durd+/emjJlivU5VTt27NA///nPHC8HBx4EbggGALhTrsPOBx98oGHDhumll17SrVu3/hzE0VFRUVF6++237V4gAABAfuQ67Li7u2vWrFl6++23derUKRmGoSpVqqh48eIFUR8AAEC+5PmmgplPIa9WrZqKFy+e44nGAAAAhSXXYefChQsKDw9XtWrV9Nxzz+n8+fOSpJdffjlXV2IBAAA8CLkOO6+//rqcnJz0yy+/yN3d3drepUsXrV271q7FAQAA5Feuz9lZv3691q1bpwoVKti0V61aVWfOnLFbYQAAAPaQ65mdlJQUmxmdTH/88YdcXFzsUhQAAIC95DrsPPvss1q4cKH1tcViUUZGht5++22bRz8AAAAUBbk+jPX2228rNDRUu3fv1s2bN/XGG2/op59+0sWLF7Vt27aCqBEAACDPcj2zU7NmTf34449q0KCBIiIilJKSog4dOmjfvn2qXLlyQdQIAACQZ7ma2bl165aaN2+uDz/8UGPHji2omgAAAOwmVzM7Tk5OOnTokCwWS0HVAwAAYFe5PozVo0cPzZkzpyBqAQAAsLtcn6B88+ZNffLJJ9qwYYPq1auX5ZlYU6dOtVtxAAAA+XVfYefHH39UcHCwihUrpkOHDunJJ5+UJB0/ftymH4e3AABAUXNfYadOnTo6f/68fHx8dObMGe3atUtlypQp6NoAAADy7b7O2SlVqpROnz4tSYqLi1NGRkaBFgUAAGAv9zWz07FjR4WEhKhcuXKyWCyqV6+eHBwcsu37888/27VAAACA/LivsPPRRx+pQ4cOOnnypF577TX9/e9/l4eHR0HXBgAAkG/3fTVWy5YtJUl79uzRoEGDCDsAAOChkOtLz+fNm1cQdQAAABSIXN9UEAAA4GFC2AEAAKZG2AEAAKZG2AEAAKaW6xOUAaAgVRq+2m5jxU1sbbexADy8mNkBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACmRtgBAACm5ljYBQAAzKfS8NV2GSduYmu7jIO/NmZ2AACAqRF2AACAqRVq2Pnf//6nNm3ayN/fXxaLRStXrrRZbxiGoqOj5e/vLzc3N4WGhuqnn36y6ZOamqqBAwfK29tbxYsX1/PPP6+zZ88+wL0AAABFWaGGnZSUFD3++OOaOXNmtusnT56sqVOnaubMmdq1a5f8/PwUERGhK1euWPsMHjxYK1as0LJly7R161ZdvXpVkZGRSk9Pf1C7AQAAirBCPUG5VatWatWqVbbrDMPQ9OnTNWrUKHXo0EGStGDBAvn6+mrJkiXq27evkpOTNWfOHC1atEjNmjWTJH366acKCAjQxo0b1aJFiwe2LwAAoGgqsufsnD59WgkJCWrevLm1zcXFRSEhIdq+fbskac+ePbp165ZNH39/fwUHB1v7ZCc1NVWXL1+2WQAAgDkV2bCTkJAgSfL19bVp9/X1ta5LSEiQs7OzSpcufdc+2ZkwYYI8PT2tS0BAgJ2rBwAARUWRDTuZLBaLzWvDMLK03elefUaMGKHk5GTrEh8fb5daAQBA0VNkw46fn58kZZmhSUxMtM72+Pn56ebNm0pKSrprn+y4uLioZMmSNgsAADCnIht2goKC5Ofnpw0bNljbbt68qS1btqhJkyaSpLp168rJycmmz/nz53Xo0CFrHwAA8NdWqFdjXb16VSdPnrS+Pn36tPbv3y8vLy9VrFhRgwcP1vjx41W1alVVrVpV48ePl7u7u7p27SpJ8vT0VFRUlIYOHaoyZcrIy8tLw4YNU61ataxXZwEAgL+2Qg07u3fvVlhYmPX1kCFDJEk9e/bU/Pnz9cYbb+j69evq16+fkpKS1LBhQ61fv14eHh7W90ybNk2Ojo7q3Lmzrl+/rvDwcM2fP18ODg4PfH8AAEDRU6hhJzQ0VIZh3HW9xWJRdHS0oqOj79rH1dVVM2bM0IwZMwqgQgAA8LArsufsAAAA2ANhBwAAmBphBwAAmBphBwAAmBphBwAAmBphBwAAmBphBwAAmBphBwAAmBphBwAAmFqh3kEZAFB4Kg1fbbex4ia2tttYgL0xswMAAEyNsAMAAEyNw1gAgIeKvQ6/meXQG4cj742ZHQAAYGqEHQAAYGqEHQAAYGqEHQAAYGqEHQAAYGqEHQAAYGqEHQAAYGqEHQAAYGqEHQAAYGqEHQAAYGo8LgIoZNz6HgAKFjM7AADA1Ag7AADA1Ag7AADA1Ag7AADA1DhBGcBfBieDA39NzOwAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTI+wAAABTK9JhJzo6WhaLxWbx8/OzrjcMQ9HR0fL395ebm5tCQ0P1008/FWLFAACgqCnSYUeSHnvsMZ0/f966HDx40Lpu8uTJmjp1qmbOnKldu3bJz89PERERunLlSiFWDAAAihLHwi7gXhwdHW1mczIZhqHp06dr1KhR6tChgyRpwYIF8vX11ZIlS9S3b9+7jpmamqrU1FTr68uXL9u/cAAAUCQU+ZmdEydOyN/fX0FBQXrxxRf1888/S5JOnz6thIQENW/e3NrXxcVFISEh2r59e45jTpgwQZ6entYlICCgQPcBAAAUniIddho2bKiFCxdq3bp1+vjjj5WQkKAmTZrowoULSkhIkCT5+vravMfX19e67m5GjBih5ORk6xIfH19g+wAAAApXkT6M1apVK+ufa9WqpcaNG6ty5cpasGCBGjVqJEmyWCw27zEMI0vbnVxcXOTi4mL/ggEAMJFKw1fbZZy4ia3tMk5eFemZnTsVL15ctWrV0okTJ6zn8dw5i5OYmJhltgcAAPx1PVRhJzU1VUeOHFG5cuUUFBQkPz8/bdiwwbr+5s2b2rJli5o0aVKIVQIAgKKkSB/GGjZsmNq0aaOKFSsqMTFR48aN0+XLl9WzZ09ZLBYNHjxY48ePV9WqVVW1alWNHz9e7u7u6tq1a2GXDgAAiogiHXbOnj2rv/3tb/rjjz9UtmxZNWrUSDt27FBgYKAk6Y033tD169fVr18/JSUlqWHDhlq/fr08PDwKuXIAAFBUFOmws2zZshzXWywWRUdHKzo6+sEUBAAAHjoP1Tk7AAAAuUXYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApkbYAQAApmaasDNr1iwFBQXJ1dVVdevW1XfffVfYJQEAgCLAFGFn+fLlGjx4sEaNGqV9+/bpmWeeUatWrfTLL78UdmkAAKCQmSLsTJ06VVFRUXr55Zf16KOPavr06QoICNDs2bMLuzQAAFDIHAu7gPy6efOm9uzZo+HDh9u0N2/eXNu3b8/2PampqUpNTbW+Tk5OliRdvnzZ7vVlpF6z21h31mevsbPbb8YumM+6IMfms35wY/NZP7ixzfJZF+TYD+v3aM9xDcPIuaPxkPv1118NSca2bdts2v/zn/8Y1apVy/Y9Y8aMMSSxsLCwsLCwmGCJj4/PMSs89DM7mSwWi81rwzCytGUaMWKEhgwZYn2dkZGhixcvqkyZMnd9T0G6fPmyAgICFB8fr5IlSxb5cRn7wY3L2A927Iex5od17Iex5od17Iex5vtlGIauXLkif3//HPs99GHH29tbDg4OSkhIsGlPTEyUr69vtu9xcXGRi4uLTVupUqUKqsT7VrJkyQL5YSmocRn7wY3L2A927Iex5od17Iex5od17Iex5vvh6el5zz4P/QnKzs7Oqlu3rjZs2GDTvmHDBjVp0qSQqgIAAEXFQz+zI0lDhgxR9+7dVa9ePTVu3FgfffSRfvnlF/3jH/8o7NIAAEAhM0XY6dKliy5cuKB///vfOn/+vIKDg/Xtt98qMDCwsEu7Ly4uLhozZkyWQ2tFdVzGfnDjMvaDHfthrPlhHfthrPlhHfthrNneLIZxr+u1AAAAHl4P/Tk7AAAAOSHsAAAAUyPsAAAAUyPsAAAAUyPsFLLt27fLwcFBLVu2tMt4Foslx6VXr175Gr9Xr17WsRwdHVWxYkW9+uqrSkpKynftiYmJ6tu3rypWrCgXFxf5+fmpRYsW+v777/M99u11Ozk56ZFHHtGwYcOUkpJit3FvX/L6fX7wwQfy8PBQWlqate3q1atycnLSM888Y9P3u+++k8Vi0fHjx/NVf7t27fL8/uwkJCRo0KBBqlKlilxdXeXr66unn35aH3zwga5dy9tzdtq0aaNmzZplu+7777+XxWLR3r1781P2Xb/LkydP5mvczLHt/TkXFMMw1KxZM7Vo0SLLulmzZsnT01O//PJLnsa+8++hr6+vIiIiNHfuXGVkZOS3dEl//vwNHDhQjzzyiFxcXBQQEKA2bdpo06ZNeR4zs+6JEyfatK9cudIud93P7ufjv//9r1xdXTV58mS7jitJsbGxslgsunTpUp7HtVgs2d7epV+/fnb5XWNvhJ1CNnfuXA0cOFBbt27N8z8gtzt//rx1mT59ukqWLGnT9u677+Z7Gy1bttT58+cVFxenTz75RF9//bX69euX73E7duyoAwcOaMGCBTp+/LhWrVql0NBQXbx4Md9jS/+v7p9//lnjxo3TrFmzNGzYMLuNe/uydOnSPI0VFhamq1evavfu3da27777Tn5+ftq1a5dNWIiNjZW/v7+qVauW732wl59//ll16tTR+vXrNX78eO3bt08bN27U66+/rq+//lobN27M07hRUVHavHmzzpw5k2Xd3Llz9cQTT+jJJ5/Mb/nZfpdBQUH5HregxMfHKyoqSv7+/nJ2dlZgYKAGDRqkCxcu5HlMi8WiefPm6YcfftCHH35obT99+rT+9a9/6d1331XFihXzPP7t/36sWbNGYWFhGjRokCIjI21Cfl7ExcWpbt262rx5syZPnqyDBw9q7dq1CgsLU//+/fM1tqurqyZNmmSX/9jdyyeffKJu3bpp5syZeuONNwp8e3kREBCgZcuW6fr169a2GzduaOnSpfn6+SgoprjPzsMqJSVFn332mXbt2qWEhATNnz9fo0ePzteYfn5+1j97enrKYrHYtNlD5qyLJFWoUEFdunTR/Pnz8zXmpUuXtHXrVsXGxiokJESSFBgYqAYNGuS3XKvb6+7atatiYmK0cuVKzZ49227j5lf16tXl7++v2NhYNWrUSNKfoaZt27aKiYnR9u3brTMcsbGxCgsLs8t27aVfv35ydHTU7t27Vbx4cWt7rVq11LFjx3s/mfguIiMj5ePjo/nz52vMmDHW9mvXrmn58uUaP358vmuX7PtdFrSff/5ZjRs3VrVq1bR06VIFBQXpp59+0j//+U+tWbNGO3bskJeXV57GDggI0LvvvqsBAwaoefPmqlSpkqKiohQeHp7v/7Hf/hmXL19eTz75pBo1aqTw8HDNnz9fL7/8cp7HzpxV2Llzp83P32OPPaY+ffrkq+5mzZrp5MmTmjBhQr5mW+5l8uTJGj16tJYsWaKOHTsW2Hby68knn9TPP/+sL7/8Ut26dZMkffnllwoICNAjjzxSyNVlxcxOIVq+fLmqV6+u6tWr66WXXtK8efPy/MugsPz8889au3atnJyc8jVOiRIlVKJECa1cuVKpqal2qi5nbm5uunXr1gPZVm6EhoYqJibG+jomJkahoaEKCQmxtt+8eVPff/99kQo7Fy5c0Pr169W/f3+bXzS3y+uUv6Ojo3r06KH58+fb/B35/PPPdfPmTes/tn8l/fv3l7Ozs9avX6+QkBBVrFhRrVq10saNG/Xrr79q1KhR+Rq/Z8+eCg8PV+/evTVz5kwdOnRIH330kZ2qt9W0aVM9/vjj+vLLL/M8xsWLF7V27dq7/vzl9/mHDg4OGj9+vGbMmKGzZ8/ma6y7GT58uN566y198803RTroZOrdu7fmzZtnfT137tx8h8qCQtgpRHPmzNFLL70k6c+p3atXr+bruPKD8s0336hEiRJyc3NT5cqVdfjwYf3rX//K15iOjo6aP3++FixYoFKlSumpp57SyJEj9eOPP9qpals7d+7UkiVLFB4enu+xMj+P25e33norz+OFhoZq27ZtSktL05UrV7Rv3z49++yzCgkJUWxsrCRpx44dun79epEKOydPnpRhGKpevbpNu7e3t/Vzyc/PSZ8+fRQXF2f9DKQ//3Ht0KGDSpcunedxb3fnd/nCCy/YZVx7u3jxotatW6d+/frJzc3NZp2fn5+6deum5cuX5/s/Tx999JEOHz6swYMH68MPP5SPj0++xstJjRo1FBcXl+f3Z/781ahRw35F3aF9+/Z64oknbGYX7WXNmjWaNGmSvvrqq7uen5YX2f371KpVK7uM3b17d23dulVxcXE6c+aMtm3bZv2dVtRwGKuQHDt2TDt37rT+T8bR0VFdunTR3Llz7fqDXhDCwsI0e/ZsXbt2TZ988omOHz+ugQMH5nvcjh07qnXr1vruu+/0/fffa+3atZo8ebI++eQTu5zslvmXPi0tTbdu3VLbtm01Y8aMfI+b+XncLq+HDzLHS0lJ0a5du5SUlKRq1arJx8dHISEh6t69u1JSUhQbG6uKFSsWyeniO2dvdu7cqYyMDHXr1i1fs3Y1atRQkyZNNHfuXIWFhenUqVP67rvvtH79+vyWbHXnd3m3GarCduLECRmGoUcffTTb9Y8++qiSkpL0+++/5yug+Pj46JVXXtHKlSvVvn37PI9zPwzDyNfJvpnBzh4nDOdk0qRJatq0qYYOHWrXcWvXrq0//vhDo0ePVv369eXh4WGXcbP79+mHH36wSyjx9vZW69attWDBAhmGodatW8vb2zvf4xYEwk4hmTNnjtLS0lS+fHlrm2EYcnJyUlJSkt3+p1oQihcvripVqkiS3nvvPYWFhWns2LH5ms3I5OrqqoiICEVERGj06NF6+eWXNWbMGLuEncy/9E5OTvL398/3obdMt38e9lClShVVqFBBMTExSkpKsp7D5Ofnp6CgIG3btk0xMTFq2rSp3bZpD1WqVJHFYtHRo0dt2jMD2Z0zEHkRFRWlAQMG6P3339e8efMUGBhol9m5TPb+LgtL5i9+Z2fnfI/l6OgoR8eC/1Vx5MiRfJ0MXrVqVVksFh05cqRAr3x79tln1aJFC40cOdKuVxyVL19eX3zxhcLCwtSyZUutXbvWLoEnu59pex6G69OnjwYMGCBJev/99+02rr1xGKsQpKWlaeHChZoyZYr2799vXQ4cOKDAwEAtXry4sEvMlTFjxuidd97RuXPn7D52zZo18315eKbMv/SBgYF2CzoFJSwsTLGxsYqNjVVoaKi1PSQkROvWrdOOHTuK1CEsSSpTpowiIiI0c+ZMu31nd+rcubMcHBy0ZMkSLViwQL179y7w/8kXRZnB8vDhw9muP3r0qMqWLZvv81QelM2bN+vgwYP5Ok/Fy8tLLVq00Pvvv5/tz19eL7POzsSJE/X1119r+/btdhtTkipWrKgtW7YoMTFRzZs31+XLl+06fkFo2bKlbt68qZs3b2Z7u4KigrBTCL755hslJSUpKipKwcHBNkunTp00Z86cwi4xV0JDQ/XYY4/l64qYCxcuqGnTpvr000/1448/6vTp0/r88881efJktW3b1o7V2l9qaqoSEhJslj/++CNfY4aFhWnr1q3av3+/dWZH+jPsfPzxx7px40aRCzvSn/dhSUtLU7169bR8+XIdOXJEx44d06effqqjR4/KwcEhX+OXKFFCXbp00ciRI3Xu3Lkidy+PnCQnJ9v852b//v15vt1EZrCcNWuWzaW/0p/3mVm8eHGR/Wwy/778+uuv2rt3r8aPH6+2bdsqMjJSPXr0yNfYs2bNUnp6uho0aKAvvvhCJ06c0JEjR/Tee++pcePGdtqDP68u7Natm10Og9+pQoUKio2N1YULF9S8eXMlJyfbfRv25ODgoCNHjujIkSP5/vtdkAg7hWDOnDlq1qyZPD09s6zr2LGj9u/fn+8bpD1oQ4YM0ccff6z4+Pg8vb9EiRJq2LChpk2bpmeffVbBwcF688039fe//10zZ860c7X2tXbtWpUrV85mefrpp/M1ZlhYmK5fv64qVarI19fX2h4SEqIrV66ocuXKCggIyG/pdle5cmXt27dPzZo104gRI/T444+rXr16mjFjhoYNG2aXQ51RUVFKSkpSs2bNiuT9PO4mNjZWderUsVnyc6uJmTNnKjU1VS1atND//vc/xcfHa+3atYqIiFC1atXyfRuLgpL596VSpUpq2bKlYmJi9N577+mrr77K9y/LoKAg7d27V2FhYRo6dKiCg4MVERGhTZs25fsWE3d66623Cuzq2fLly2vLli26dOmSIiIi7DorVRBKliypkiVLFnYZObIYD9u1zgAASX/eRC86Olpr165VYmKiDMNQhw4dtGjRIrm7uxd2eUCRQdgBAJMYM2aMpk6dqvXr19v1sA3wsCPsAICJzJs3T8nJyXrttddUrBhnKgASYQcAAJgcsR8AAJgaYQcAAJgaYQcAAJgaYQcAAJgaYQcAAJgaYQfAQyE0NFSDBw8u7DLuyWKxaOXKlXddHxcXJ4vFov379z+wmoC/Op56DsDutm/frmeeeUYRERFau3atXcb88ssvi/wDXCXp/PnzKl26dGGXAeA2zOwAsLu5c+dq4MCB2rp1a54fdnknLy8veXh42GWsguTn5ycXF5fCLgPAbQg7AOwqJSVFn332mV599VVFRkZq/vz5NutjY2NlsVi0bt061alTR25ubmratKkSExO1Zs0aPfrooypZsqT+9re/6dq1a9b33XkYq1KlSho/frz69OkjDw8PVaxYUR999JHNtg4ePKimTZvKzc1NZcqU0SuvvKKrV6/etfb09HRFRUUpKChIbm5uql69ut59990s/ebOnavHHntMLi4uKleunAYMGGBdd+dhrJ07d6pOnTpydXVVvXr1tG/fvvv8JAHYC2EHgF0tX75c1atXV/Xq1fXSSy9p3rx52T4dOjo6WjNnztT27dsVHx+vzp07a/r06VqyZIlWr16tDRs2aMaMGTlua8qUKdYA0a9fP7366qs6evSoJOnatWtq2bKlSpcurV27dunzzz/Xxo0bbYLJnTIyMlShQgV99tlnOnz4sEaPHq2RI0fqs88+s/aZPXu2+vfvr1deeUUHDx7UqlWrVKVKlWzHS0lJUWRkpKpXr649e/YoOjpaw4YNu5+PEYA9GQBgR02aNDGmT59uGIZh3Lp1y/D29jY2bNhgXR8TE2NIMjZu3GhtmzBhgiHJOHXqlLWtb9++RosWLayvQ0JCjEGDBllfBwYGGi+99JL1dUZGhuHj42PMnj3bMAzD+Oijj4zSpUsbV69etfZZvXq1UaxYMSMhIeG+96dfv35Gx44dra/9/f2NUaNG3bW/JGPFihWGYRjGhx9+aHh5eRkpKSnW9bNnzzYkGfv27bvvGgDkDzM7AOzm2LFj2rlzp1588UVJkqOjo7p06aK5c+dm6Vu7dm3rn319feXu7q5HHnnEpi0xMTHH7d0+hsVikZ+fn/U9R44c0eOPP67ixYtb+zz11FPKyMjQsWPH7jrmBx98oHr16qls2bIqUaKEPv74Y+t5R4mJiTp37pzCw8NzrCtTZg3u7u7WNp5GDjx4XI0FwG7mzJmjtLQ0lS9f3tpmGIacnJyUlJRkc5XS7VdWWSyWLFdaWSwWZWRk5Li9nN5jGIYsFku277tb+2effabXX39dU6ZMUePGjeXh4aG3335bP/zwgyTJzc0tx3ruZPCcZaBIYGYHgF2kpaVp4cKFmjJlivbv329dDhw4oMDAQC1evPiB1lOzZk3t379fKSkp1rZt27apWLFiqlatWrbv+e6779SkSRP169dPderUUZUqVXTq1Cnreg8PD1WqVEmbNm267xoOHDig69evW9t27NiRxz0CkFeEHQB28c033ygpKUlRUVEKDg62WTp16qQ5c+Y80Hq6desmV1dX9ezZU4cOHVJMTIwGDhyo7t27y9fXN9v3VKlSRbt379a6det0/Phxvfnmm9q1a5dNn+joaE2ZMkXvvfeeTpw4ob179971ROquXbuqWLFiioqK0uHDh/Xtt9/qnXfesfu+AsgZYQeAXcyZM0fNmjWTp6dnlnUdO3bU/v37tXfv3gdWj7u7u9atW6eLFy+qfv366tSpk8LDwzVz5sy7vucf//iHOnTooC5duqhhw4a6cOGC+vXrZ9OnZ8+emj59umbNmqXHHntMkZGROnHiRLbjlShRQl9//bUOHz6sOnXqaNSoUZo0aZJd9xPAvVkMDioDAAATY2YHAACYGmEHAACYGmEHAACYGmEHAACYGmEHAACYGmEHAACYGmEHAACYGmEHAACYGmEHAACYGmEHAACYGmEHAACY2v8HZZyMf/ZsFD0AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# visualize all 20 amino acid occurrences in the form of a histogram\n",
    "\n",
    "pylab.bar(common_amino.keys(),common_amino.values())\n",
    "\n",
    "pylab.title(\"%i protein sequences\\n frquency %i to %i\" \n",
    "            % (len(common_amino.values()), \n",
    "               min(common_amino.values()), \n",
    "               max(common_amino.values())))\n",
    "\n",
    "pylab.xlabel(\"Amino acid\")\n",
    "pylab.ylabel(\"frequency\")\n",
    "pylab.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Since stop codon * signifies end of a protein we can split the sequence using ( * )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATR',\n",
       " 'SPEWIPGSTSR',\n",
       " 'AWVFPVRLSLVLGQYQADGSTLACLTLNKAAI',\n",
       " 'KQVSVCSHLF',\n",
       " 'SPPGHSVSWLEVSVSSPWNCQNPLTKGQP',\n",
       " 'VKNLWQFGARTGTRET',\n",
       " 'ESTRLSEGSKGGRDQPRRRKTRCQRAERDVKVREAFGTDSPKVCVAIELFSYPQGG',\n",
       " 'KYSGRWARDTPSCQGKKQMN',\n",
       " 'KKLGYGPAERKSIC',\n",
       " 'SI']"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_list = [str(i) for i in protein_seq.split('*')]\n",
    "protein_list[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert sequences to dataframe\n",
    "large_proteins = [x for x in protein_list if len(x)> 10]\n",
    "df = pd.DataFrame({'protein_seq':large_proteins})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPEWIPGSTSR</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AWVFPVRLSLVLGQYQADGSTLACLTLNKAAI</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPPGHSVSWLEVSVSSPWNCQNPLTKGQP</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VKNLWQFGARTGTRET</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ESTRLSEGSKGGRDQPRRRKTRCQRAERDVKVREAFGTDSPKVCVA...</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         protein_seq  length\n",
       "0                                        SPEWIPGSTSR      11\n",
       "1                   AWVFPVRLSLVLGQYQADGSTLACLTLNKAAI      32\n",
       "2                      SPPGHSVSWLEVSVSSPWNCQNPLTKGQP      29\n",
       "3                                   VKNLWQFGARTGTRET      16\n",
       "4  ESTRLSEGSKGGRDQPRRRKTRCQRAERDVKVREAFGTDSPKVCVA...      56"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a column with sequence lengths\n",
    "\n",
    "df['length'] = df['protein_seq'].apply(len)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>GSFTSCIRPVSMASPGLHLLIGILLLSVLQTWCGLYVTVFYGVPAW...</td>\n",
       "      <td>902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>RLGRMDPRERAPPGNSGEETIGEAFEWLETTVTQINREAVNHLPRE...</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>MAAPPQEEAPQREPWDEWVIEVLEDIKQEALKHFDPRLLHALGNYI...</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>IWHPTASGKTNSKYMSQMSTKRRSNTWASRCISRDMADGLHSFRGK...</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SIRKRNKSNINDWGDSNKYFRQKYITAIRNDFKLSIGKSRANTSTL...</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>QEEANHKVRWRTERKLAEIAGTFQQGDGQWVETGRGVWERPIIFCI...</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ESTRLSEGSKGGRDQPRRRKTRCQRAERDVKVREAFGTDSPKVCVA...</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EGNNRRGHQLLDALIVGNQATLPEIAGHQEERGAGNVERKDTCKQT...</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>RSQITGQCSTENRQGEPSNMGTITNISPTSRERNLGIMVDRLLASHVDT</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NHQSTGRARSNKNGFRRLRRQSKYGHRLTVCHRYFIRTAYRIRI</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          protein_seq  length\n",
       "56  GSFTSCIRPVSMASPGLHLLIGILLLSVLQTWCGLYVTVFYGVPAW...     902\n",
       "53  RLGRMDPRERAPPGNSGEETIGEAFEWLETTVTQINREAVNHLPRE...     115\n",
       "54  MAAPPQEEAPQREPWDEWVIEVLEDIKQEALKHFDPRLLHALGNYI...     100\n",
       "41  IWHPTASGKTNSKYMSQMSTKRRSNTWASRCISRDMADGLHSFRGK...      98\n",
       "21  SIRKRNKSNINDWGDSNKYFRQKYITAIRNDFKLSIGKSRANTSTL...      90\n",
       "63  QEEANHKVRWRTERKLAEIAGTFQQGDGQWVETGRGVWERPIIFCI...      63\n",
       "4   ESTRLSEGSKGGRDQPRRRKTRCQRAERDVKVREAFGTDSPKVCVA...      56\n",
       "16  EGNNRRGHQLLDALIVGNQATLPEIAGHQEERGAGNVERKDTCKQT...      54\n",
       "33  RSQITGQCSTENRQGEPSNMGTITNISPTSRERNLGIMVDRLLASHVDT      49\n",
       "36       NHQSTGRARSNKNGFRRLRRQSKYGHRLTVCHRYFIRTAYRIRI      44"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sort sequence data\n",
    "\n",
    "df.sort_values(by = ['length'], ascending = False)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GSFTSCIRPVSMASPGLHLLIGILLLSVLQTWCGLYVTVFYGVPAWKNATVPLFCATKNRDTWGTTQCLPDNNDYVEIPLNVTEAFDAWNNTVTEQAIEDVWNLFETSIKPCVKLTPLCVAMRCNKTETEKWGLTNKQNTTNTTSTTTTSTPTTTKEKEIKGNETCINDNNCAGLETEPMIGCKFNMSGLKRDEEKEYNETWYSQDLVCDQNYNSTNISRCYMNHCNTSVIQESCDKHYWDSLRFRYCAPPGFALLRCNDTSYSGFQPNCSKVVASSCTRMMETQTSTWFGLNGTRAENRTYIYWHGRDNRTIISLNKYYNLTMRCRRPGNKTVRPITIMSGLVFHSQPINERPKQAWCWFEGEWKKAIQEVKDTLATHPRYTGTNDTKKINLVAPEGGDPEVTFMWTNCRGEFLYCKMNWFLNWVEDRNITHGRWSTQKPAEKQKRNYVPCHIRQIINTWHKVGKNVYLPPREGNLTCNSSVTSIIANIDWISDNETNITMSAEVAELYRLELGDYKLVEITPIGLAPTSVKRYSSATPRNKRGVFVLGFLGFLATAGSAMGAASLTLTAQSRTLLAGIVQQQQQLLDAVKRQQELLRLTVWGTKNLQTRVTAIEKYLKDQAQLNSWGCAFRQVCHTTVPWPNDTLQPNWDNMTWQEWERKVDFLTENITELLEQAQIQQEKNMYELQKLNSWDVFGNWFDLSSWITYIQYGVYLVVGVIGLRISIYIVQMLLRLREGYRPVFSSPPSYRQQIHIRRDQELPDGEDREEDGGEKGGNRSWPWQIEYIHFLIRQLIRLLTWLYSNCRDLIYRSFQTLHQLTQWILRRISAAATATRDFIRTEASYISYGWQYFLEALQAAMRAAGETLASAGGELWATLGRIGRGLLAVPRRIRQGLELALN'"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# let's take a single protein from the table\n",
    "\n",
    "one_large_protein = df.nlargest(1,'length')\n",
    "single_prot = one_large_protein.iloc[0,0]\n",
    "single_prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein sequence written to single_protein_sequence.fasta in FASTA format\n"
     ]
    }
   ],
   "source": [
    "# Define the filename\n",
    "fasta_filename = \"single_protein_sequence.fasta\"\n",
    "\n",
    "# Write the sequence in FASTA format\n",
    "with open(fasta_filename, 'w') as file:\n",
    "    # Write the header (starting with '>')\n",
    "    file.write(\">single_large_protein\\n\")\n",
    "    \n",
    "    # Write the sequence\n",
    "    file.write(single_prot)\n",
    "\n",
    "print(f\"Protein sequence written to {fasta_filename} in FASTA format\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Basic Local Alignment Using NCBI-BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSFTSCIRPVSMASPGLHLLIGILLLSVLQTWCGLYVTVFYGVPAWKNATVPLFCATKNRDTWGTTQCLPDNNDYVEIPLNVTEAFDAWNNTVTEQAIEDVWNLFETSIKPCVKLTPLCVAMRCNKTETEKWGLTNKQNTTNTTSTTTTSTPTTTKEKEIKGNETCINDNNCAGLETEPMIGCKFNMSGLKRDEEKEYNETWYSQDLVCDQNYNSTNISRCYMNHCNTSVIQESCDKHYWDSLRFRYCAPPGFALLRCNDTSYSGFQPNCSKVVASSCTRMMETQTSTWFGLNGTRAENRTYIYWHGRDNRTIISLNKYYNLTMRCRRPGNKTVRPITIMSGLVFHSQPINERPKQAWCWFEGEWKKAIQEVKDTLATHPRYTGTNDTKKINLVAPEGGDPEVTFMWTNCRGEFLYCKMNWFLNWVEDRNITHGRWSTQKPAEKQKRNYVPCHIRQIINTWHKVGKNVYLPPREGNLTCNSSVTSIIANIDWISDNETNITMSAEVAELYRLELGDYKLVEITPIGLAPTSVKRYSSATPRNKRGVFVLGFLGFLATAGSAMGAASLTLTAQSRTLLAGIVQQQQQLLDAVKRQQELLRLTVWGTKNLQTRVTAIEKYLKDQAQLNSWGCAFRQVCHTTVPWPNDTLQPNWDNMTWQEWERKVDFLTENITELLEQAQIQQEKNMYELQKLNSWDVFGNWFDLSSWITYIQYGVYLVVGVIGLRISIYIVQMLLRLREGYRPVFSSPPSYRQQIHIRRDQELPDGEDREEDGGEKGGNRSWPWQIEYIHFLIRQLIRLLTWLYSNCRDLIYRSFQTLHQLTQWILRRISAAATATRDFIRTEASYISYGWQYFLEALQAAMRAAGETLASAGGELWATLGRIGRGLLAVPRRIRQGLELALN\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Provide the correct file path\n",
    "file_path = \"single_protein_sequence.fasta\"\n",
    "\n",
    "# Read the sequence from the FASTA file\n",
    "read = SeqIO.read(file_path, \"fasta\")\n",
    "\n",
    "# Print the sequence\n",
    "print(read.seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: total: 0 ns\n",
      "Wall time: 1min 4s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# based on the server load this query might take 2-3 minutes to run\n",
    "\n",
    "result_handle = NCBIWWW.qblast(\"blastp\",\"pdb\",read.seq)\n",
    "\n",
    "blast_qresult = SearchIO.read(result_handle, \"blast-xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Program: blastp (2.16.0+)\n",
      "  Query: unnamed (902)\n",
      "         protein product\n",
      " Target: pdb\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  pdb|7T2P|A  Chain A, Envelope glycoprotein gp120 [Simia...\n",
      "            1      1  pdb|8DUA|C  Chain C, Envelope glycoprotein gp120 [Simia...\n",
      "            2      1  pdb|8DVD|E  Chain E, Envelope glycoprotein gp160 [Simia...\n",
      "            3      2  pdb|6TYB|G  Chain G, Envelope glycoprotein gp160 [Simia...\n",
      "            4      2  pdb|2BF1|A  Chain A, EXTERIOR MEMBRANE GLYCOPROTEIN GP1...\n"
     ]
    }
   ],
   "source": [
    "print(blast_qresult[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence ID:pdb|7T2P|A\n",
      "description:Chain A, Envelope glycoprotein gp120 [Simian immunodeficiency virus]\n",
      "E value:    0.0 \n",
      "Bit Score:  875.159\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#fetch the id, description, evalue, bitscore & alignment of first hit\n",
    "\n",
    "seqid = blast_qresult[0]\n",
    "\n",
    "details = seqid[0]\n",
    "\n",
    "print(f\"\\\n",
    "Sequence ID:{seqid.id}\\n\\\n",
    "description:{seqid.description}\\n\\\n",
    "E value:    {details.evalue} \\n\\\n",
    "Bit Score:  {details.bitscore}\\n\\\n",
    "\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alignment:\n",
      "Alignment with 2 rows and 527 columns\n",
      "MASPGLHLLIGILLLSVLQTWCGLYVTVFYGVPAWKNATVPLFC...SSA unnamed\n",
      "MGCLGNQLLIAILLLSVYGIYCTLYVTVFYGVPAWRNATIPLFC...TTG pdb|7T2P|A\n"
     ]
    }
   ],
   "source": [
    "print(f\"alignment:\\n{details.aln}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Reading PDB file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retreiving PDB Structure From RCSB PDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'pdb|7T2P|A'"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqid.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'7T2P'"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# split seqid\n",
    "\n",
    "seqid.id.split('|')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File downloaded and saved at Sequence_data\\7T2P.pdb\n"
     ]
    }
   ],
   "source": [
    "import urllib.request\n",
    "import os\n",
    "\n",
    "# Define the directory and file path\n",
    "directory = \"Sequence_data\"\n",
    "file_path = os.path.join(directory, \"7T2P.pdb\")\n",
    "\n",
    "# Create the directory if it doesn't exist\n",
    "if not os.path.exists(directory):\n",
    "    os.makedirs(directory)\n",
    "\n",
    "# Download the PDB file and save it to the specified directory\n",
    "urllib.request.urlretrieve('https://files.rcsb.org/download/7T2P.pdb', file_path)\n",
    "\n",
    "print(f\"File downloaded and saved at {file_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.1 Reading PDB file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Structure id=7T2P>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\nirmi\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8643.\n",
      "  warnings.warn(\n",
      "C:\\Users\\nirmi\\anaconda3\\Lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:100: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 8783.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "from Bio.PDB import PDBParser\n",
    "\n",
    "# Suppress PDBConstructionWarnings\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "\n",
    "# Initialize PDB parser\n",
    "parser = PDBParser()\n",
    "\n",
    "# Parse the PDB structure\n",
    "structure = parser.get_structure(\"7T2P\", \"Sequence_data/7T2P.pdb\")\n",
    "\n",
    "# Print the structure\n",
    "print(structure)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.2 Identifying the number of chains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chainid: A\n",
      "chainid: B\n",
      "chainid: H\n",
      "chainid: L\n",
      "chainid: C\n",
      "chainid: D\n",
      "chainid: E\n",
      "chainid: F\n",
      "chainid: G\n",
      "chainid: I\n",
      "chainid: J\n",
      "chainid: K\n",
      "chainid: M\n",
      "chainid: N\n",
      "chainid: O\n",
      "chainid: P\n"
     ]
    }
   ],
   "source": [
    "for chain in structure[0]:\n",
    "    print(f\"chainid: {chain.id}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Visualizing SARS-CoV-19 Protein structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1. nglview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6aa03d2edb544567869c45fc4c4258f4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nv.demo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "93176d9294a64acaaaa1e229610d2ad2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "view = nv.show_biopython(structure)\n",
    "view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "41eeb5b539a1472a84aab3a3d01f48c2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Image(value=b'', width='99%')"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "view.render_image()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.2. nglview GUI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f1a975a963144fc2b2853829a448dbf6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bcbb810cc3c641e09c2c89e4e11a2025",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Tab(children=(Box(children=(Box(children=(Box(children=(Label(value='step'), IntSlider(value=1, min=-100)), la"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#GUI\n",
    "view = nv.show_biopython(structure, gui=True)\n",
    "view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a76d7208f01a4837a10cf0329746ff5c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Image(value=b'', width='99%')"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "view.render_image()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
